File: mesh-control-python-server/provision.py
import time

from flask import Blueprint, current_app, jsonify, make_response, request
from utils.process import write_to_meshctl



provision_bp = Blueprint('provision_bp', __name__)


@provision_bp.route('/provision', methods=['GET', 'POST', 'DELETE'])
def handle_config():
    if request.method == 'GET':
        
        query_params = request.args.getlist('query')

        if current_app.config['PROVISION']["SCAN_ACTIVE"]:
            if current_app.config['PROVISION']["USE_FAILBACK_SCAN"]:
                failback_scan()
            else:
                normal_scan()
        if current_app.config['PROVISION']["PROVISION_ACTIVE"]:
            node_provision_output()
            stop_node_provision()

        response_value = {}

        if not query_params:
            return jsonify(current_app.config['PROVISION']), 200
        
        for query in query_params:
            if query not in current_app.config['PROVISION']:
                return jsonify({query: "Invalid query parameter"}), 400
            response_value[query] = current_app.config['PROVISION'].get(query)


        response = make_response(response_value)
        response.headers.add('Access-Control-Allow-Origin', '*')
        return response, 200

    elif request.method == 'POST':
        req_data = request.get_json()
        discovery_status = req_data.get('discovery')
        failback_scan_status = req_data.get('failback_scan')
        provision_node = req_data.get('provision_node')

        if discovery_status is not None:
            if discovery_status == "any":
                discovery_status = not current_app.config['PROVISION']['SCAN_ACTIVE']
            scan_unprovisioned(discovery_status)
            response = {
                "status": "success",
                "message": f"Discovery status is set to {discovery_status}"
            }
        elif failback_scan_status is not None:
            if failback_scan_status == "any":
                current_app.config['PROVISION']['USE_FAILBACK_SCAN'] = not current_app.config['PROVISION']['USE_FAILBACK_SCAN']
            else:
                current_app.config['PROVISION']['USE_FAILBACK_SCAN'] = failback_scan_status
            restart_scan()
            response = {
                "status": "success",
                "message": f"Failback scan status is set to {current_app.config['PROVISION']['USE_FAILBACK_SCAN']}",
                "USE_FAILBACK_SCAN": current_app.config['PROVISION']['USE_FAILBACK_SCAN']
            }            
        elif provision_node:
            start_node_provision(provision_node)
            response = {
                "status": "success",
                "message": f"Started provisioning for node {provision_node}"
            }
        else:
            response = {
                "status": "fail",
                "message": "Provision request is empty or not found"
            }              
        return jsonify(response), 201
    elif request.method == "DELETE":
        address = request.args.get('address')
        unprovisioned = request.args.get('reset')
        response_value = {}

        if unprovisioned:
            current_app.config['PROVISION']["UNPROVISIONED_NODES"].clear()
            restart_scan()
            response_value = {
                "status": "success",
                "message": "Deleted unprovisioned nodes list",
                "UNPROVISIONED_NODES":current_app.config['PROVISION']['UNPROVISIONED_NODES']
            }
        else:
            return jsonify({"REMOVE_METHODS":["unprovisioned"]}), 421
        
        response = make_response(response_value)
        response.headers.add('Access-Control-Allow-Origin', '*')
        return response, 200
            
def restart_scan():
    write_to_meshctl("discover-unprovisioned off")
    time.sleep(0.1)
    current_app.config['TERMINAL_OUTPUT'].clear()
    write_to_meshctl("discover-unprovisioned on")

def start_node_provision(node):
    scan_unprovisioned(False)
    current_app.config['PROVISION']['PROVISION_ACTIVE'] = True
    current_app.config['PROVISION']["UNPROVISIONED_NODES"].clear()
    current_app.config['PROVISION']['PROVISION_START_TIME'] = time.time()
    write_to_meshctl(f"provision {node}")

def stop_node_provision():
    error_arr = ["Failed to connect:","Stale services? Remove device and re-discover","Services resolved no","Could not find device proxy"]
    out = "".join(current_app.config['PROVISION']['PROVISION_OUTPUT'])
    if time.time() - current_app.config['PROVISION']['PROVISION_START_TIME'] >= 60:
        current_app.config['PROVISION']['PROVISION_ACTIVE'] = False
        current_app.config['PROVISION']['PROVISION_STATUS'] = "Provision failed on timeout!"
    for error in error_arr:
        if error in out:
            current_app.config['PROVISION']['PROVISION_ACTIVE'] = False
            current_app.config['PROVISION']['PROVISION_STATUS'] = "Provision failed on error!"
    if "Provision success." in out:
        time.sleep(5)
        current_app.config['PROVISION']['PROVISION_ACTIVE'] = False
        current_app.config['PROVISION']['PROVISION_STATUS'] = "Provision success!"

def node_provision_output():
    if current_app.config['PROVISION']['PROVISION_ACTIVE'] == True:
        current_app.config['PROVISION']['PROVISION_OUTPUT'] = current_app.config['TERMINAL_OUTPUT'].copy()

def scan_unprovisioned(state):
    current_app.config['TERMINAL_OUTPUT'].clear()
    current_app.config['PROVISION']['SCAN_ACTIVE'] = state
    write_to_meshctl(f"discover-unprovisioned {'on' if state else 'off'}\n")

def normal_scan():
    if "[NEW]" in "".join(current_app.config['TERMINAL_OUTPUT']):
        address, name, OOB, UUID = "", "", "", ""
        for entry in current_app.config['TERMINAL_OUTPUT']:
            if "Device UUID:" in entry:
                UUID = entry.split(" ")[-1]
            if "OOB:" in entry:
                OOB = entry.split(" ")[-1]
            if "[NEW]" in entry:
                str_arr = entry.split(" ")
                start_index = str_arr.index("Device")
                address = str_arr[start_index+1]
                name = ""
                for i in range(start_index+2, len(str_arr)):
                    name += f"{str_arr[i]} "
                name = name[:-1]
        if UUID == "":
            return
        print(f"normal_scan() - Found node: Address: {address} | Name: {name} | OOB: {OOB} | UUID: {UUID}")
        current_app.config['PROVISION']["UNPROVISIONED_NODES"][UUID] = {
                "name": name,
                "OOB": OOB,
                "address": address
            }

def failback_scan():
    if "OOB:" in "".join(current_app.config['TERMINAL_OUTPUT']):
        OOB, UUID = "", ""
        for entry in current_app.config['TERMINAL_OUTPUT']:
            if "Device UUID:" in entry:
                UUID = entry.split(" ")[-1]
            if "OOB:" in entry:
                OOB = entry.split(" ")[-1]
        print(f"failback_scan() - Found node: OOB: {OOB} | UUID: {UUID}")
        current_app.config['PROVISION']["UNPROVISIONED_NODES"][UUID] = {
            "OOB": OOB
        }

File: mesh-control-client/pages/api/provision.ts
import { NextApiRequest, NextApiResponse } from "next";
import {delay} from "./utils/common"
import { updateConfig } from "./utils/updateData";
import { startProcess } from "./utils/process";
import Debug from "./utils/debug";
import { getUuidInfo, isValidRawUuid, isValidStandardUuid, rawUuidToStandard, standardUuidToRaw } from "./utils/uuid";
import { debug } from "console";

export default async function handler(request: NextApiRequest, response: NextApiResponse) {
  response.setHeader('Access-Control-Allow-Origin', '*');
  response.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
  response.setHeader('Access-Control-Allow-Headers', 'Content-Type');

      scan_unprovisioned()
  if (request.method === 'GET') {
    let query = request.query['query'];
    let returnObj: any = {}

    if (query){
      if (typeof query == "object"){
          query.forEach(element => {
            returnObj[element] = global.DATA.PROVISION[element as keyof typeof global.DATA.PROVISION]
    });
      }
      else{
        returnObj[query] = global.DATA.PROVISION[query as keyof typeof global.DATA.PROVISION]
      }

    
    return response.status(200).json(returnObj)
     };
    
  }

  if (request.method === 'POST') {
    
    const { discovery, failback_scan_status, provision_node } = request.body;
    
    if (discovery != undefined){
      global.DATA.PROVISION.UNPROVISIONED_NODES = {}
      global.DATA.PROVISION.SCAN_ACTIVE = discovery
      if (discovery){
        Debug.log("Starting discovery", "INFO", "Provision");
        if (global.DATA.TERMINAL_SESSIONS.MESHCTL.STATUS == false){
          Debug.log("Meshctl process not running, starting for discovery", "WARNING", "Provision");
          startProcess("MESHCTL")
          await delay(500)
        }

        if (global.DATA.CONTROLLER.POWER == false){
          Debug.log("Controller power is off, turning on", "WARNING", "Provision");
          global.DATA.TERMINAL_SESSIONS.MESHCTL.PROCESS?.stdin?.write("power on\n")
          await delay(500)
        }
        global.DATA.TERMINAL_SESSIONS.MESHCTL.PROCESS?.stdin?.write("discover-unprovisioned on\n")
        global.DATA.TERMINAL_SESSIONS.MESHCTL.LOCK = true
      }  
      else{
        Debug.log("Stopping discovery", "INFO", "Provision");
        global.DATA.TERMINAL_SESSIONS.MESHCTL.PROCESS?.stdin?.write("discover-unprovisioned off\n")
        global.DATA.TERMINAL_SESSIONS.MESHCTL.LOCK = false
      }
      return response.status(200).json({"status": "success", "message": `Discovery status is set to ${discovery}` });
    }
    if (provision_node != undefined){
      Debug.log(`Trying to provision node ${provision_node}`, "INFO", "Provision");
      //Debug.fakeProcess("provision", failback_scan_status)
      provision(provision_node)
      
      return response.status(201).json({
                "status": "success",
                "message": `Started provisioning for node ${provision_node}`
            });
    }

  }
}

function stop_provision(error = false){
    if (error){
      Debug.log("Stopping provision process with error!", "ERROR", "Provision");
    }
    else {
      Debug.log("Stopping provision process!", "INFO", "Provision");
    }
    global.DATA.TERMINAL_SESSIONS.MESHCTL.LOCK = false
  global.DATA.PROVISION.PROCESS.STATUS = false
  global.DATA.PROVISION.PROCESS.ERROR = error
  global.DATA.PROVISION.PROCESS.PROGRESS = 100
  global.DATA.PROVISION.UNPROVISIONED_NODES = {}
  updateConfig()
}

async function update_provision(){

    let i = 0
    const maxAttemps = 100
    global.DATA.PROVISION.PROCESS.PROGRESS = 0
    while (global.DATA.TERMINAL_SESSIONS.MESHCTL.OUTPUT.filter(str => str.includes("Composition data for node")).length < 1){
      if (global.DATA.TERMINAL_SESSIONS.MESHCTL.OUTPUT.filter(str => str.includes("Failed to connect")).length >= 1){
        Debug.log("Got 'Failed to connect' for provisioning..", "ERROR", "Provision");
        global.DATA.PROVISION.PROCESS.LOGS.push("Got 'Failed to connect' for provisioning!")
        stop_provision(true)
        return  
      }
      if (global.DATA.TERMINAL_SESSIONS.MESHCTL.OUTPUT.filter(str => str.includes("Could not find device proxy")).length >= 1){
        Debug.log("Got 'Could not find device proxy' for provisioning..", "ERROR", "Provision");
        global.DATA.PROVISION.PROCESS.LOGS.push("Got 'Could not find device proxy' for provisioning!")
        stop_provision(true)
        return  
      }
      global.DATA.PROVISION.PROCESS.LOGS = global.DATA.TERMINAL_SESSIONS.MESHCTL.OUTPUT
      await delay(500)
      i++;
      global.DATA.PROVISION.PROCESS.PROGRESS = i
      if (i > maxAttemps){
        Debug.log("Got time limit for provisioning..", "ERROR", "Provision");
        global.DATA.PROVISION.PROCESS.LOGS.push("Got time limit for provisioning!")
        stop_provision(true)
        return
      }
    
  }
  Debug.log("Provisioned node!", "INFO", "Provision");
  await delay(500)
  global.DATA.PROVISION.PROCESS.LOGS = global.DATA.TERMINAL_SESSIONS.MESHCTL.OUTPUT
  global.DATA.PROVISION.PROCESS.LOGS.push("Succesfully provisioned node!")
  global.DATA.TERMINAL_SESSIONS.MESHCTL.PROCESS?.stdin?.write("disconnect\n")
  stop_provision()
}

async function provision(node: string){
  if (global.DATA.PROVISION.PROCESS.STATUS){
    Debug.log("Already provisioning, returning..", "ERROR", "Provision");
    return
  }
  global.DATA.TERMINAL_SESSIONS.MESHCTL.PROCESS?.stdin?.write("discover-unprovisioned off\n")
  global.DATA.PROVISION.SCAN_ACTIVE = false

  global.DATA.TERMINAL_SESSIONS.MESHCTL.LOCK = true
  global.DATA.TERMINAL_SESSIONS.MESHCTL.OUTPUT = []

  global.DATA.PROVISION.PROCESS.STATUS = true
  global.DATA.PROVISION.PROCESS.ERROR = false
  global.DATA.PROVISION.PROCESS.START_TIME = new Date().getUTCDate()

  await delay(500)
  
  global.DATA.TERMINAL_SESSIONS.MESHCTL.PROCESS?.stdin?.write(`provision ${standardUuidToRaw(node)}\n`)

  update_provision()

}

function scan_unprovisioned(){
  if (global.DATA.PROVISION.SCAN_ACTIVE){
    let UUID, OOB;
    let name = "";
    let address = ""
    for (let i = 0; i < global.DATA.TERMINAL_SESSIONS.MESHCTL.OUTPUT.length; i++){
      if (global.DATA.TERMINAL_SESSIONS.MESHCTL.OUTPUT[i].includes("OOB")){
        Debug.log("Found node! (without name and address)", "INFO", "Provision");
        let data = global.DATA.TERMINAL_SESSIONS.MESHCTL.OUTPUT[i-1].split(" ")
        UUID = UUID = data[data.length-1]
        data = global.DATA.TERMINAL_SESSIONS.MESHCTL.OUTPUT[i].split(" ")
        OOB = data[data.length-1]
        if (isValidRawUuid(UUID)){
          UUID = rawUuidToStandard(UUID)
        }
        else{
          Debug.log("Invalid UUID for node, skipping..", "WARNING", "Provision");
          continue
        }
        global.DATA.PROVISION.UNPROVISIONED_NODES[UUID as keyof typeof global.DATA.PROVISION.UNPROVISIONED_NODES] = {
                "name": UUID,
                "OOB": OOB,
                "address": address
    }      
      }
      
      if (global.DATA.TERMINAL_SESSIONS.MESHCTL.OUTPUT[i].includes("NEW")){
        Debug.log("Found node!", "INFO", "Provision");
        let data = global.DATA.TERMINAL_SESSIONS.MESHCTL.OUTPUT[i-2].split(" ")
        UUID = data[data.length-1]
        data = global.DATA.TERMINAL_SESSIONS.MESHCTL.OUTPUT[i-1].split(" ")
        OOB = data[data.length-1]
        data = global.DATA.TERMINAL_SESSIONS.MESHCTL.OUTPUT[i].split(" ")
        const deviceIndex = data.indexOf("Device")
        address = data[deviceIndex+1]
        for (let i = deviceIndex+2; i < data.length; i++){
          if (i == data.length){
            name += `${data[i]}`
          }
          else{
            name += `${data[i]} `
          }
          
        }
        if (isValidRawUuid(UUID)){
          UUID = rawUuidToStandard(UUID)
        }
        else{
          Debug.log("Invalid UUID for node, skipping..", "WARNING", "Provision");
          continue
        }
        global.DATA.PROVISION.UNPROVISIONED_NODES[UUID as keyof typeof global.DATA.PROVISION.UNPROVISIONED_NODES] = {
                "name": name,
                "OOB": OOB,
                "address": address
    }
        Debug.log(`Added node ${name} (${address}) to nodes list!`, "INFO", "Provision");
        global.DATA.TERMINAL_SESSIONS.MESHCTL.OUTPUT = []
      }
    }
  }
}